Troupon.Catalog.Core.Domain/Dtos/DealDetailsDto.cs

```diff
@@ -0,0 +1,34 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Troupon.Catalog.Core.Domain.Entities.Common;
+using Troupon.Catalog.Core.Domain.Enums;
+
+namespace Troupon.Catalog.Core.Domain.Dtos
+{
+  public record DealDetailsDto
```

Comment:
What are the advantages of using a record here instead of a class? For records I'm expecting a:

`public record Employee(string FirstName, string LastName);`

type of definition (I believe it's the main thing with the new record reference type)

Reply:
I took this approach to make the code easier to be tested (Since that way we can create a DTO with a empty constructor). In this particular case, I decided to use `record` for the concept of it. Since a DTO in this case should be immutable, I thought that a `record` would fit better than a `class`. 

Syntactically speaking, I don't see other practical difference in this case to use `record` or `class`, since we are not using the syntax sugar of defining the properties through the record constructor. 

From my perspective the idea of a immutable record fits better than a class, but this can be changed easily, let me know what are your thoughts on this

Reply:
Records are not strictly immutable, it depends how you define its properties. Same for classes.

e.g.:

```
var e1 = new Employee1("Martin Paquin");
// Error! Immutable 
// e1.Name = "Pedro Ferreira";

var e2 = new Employee2 { Name = "Martin Paquin" };
// Success! records are not strictly immutable, it depends on how you define them
e2.Name = "Pedro Ferreira";

var e3 = new Employee3 { Name = "Martin Paquin" };
// Error! Immutable, even if it's a class
//e3.Name = "Pedro Ferreira";

public record Employee1(string Name);

public record Employee2
{
    public string Name { get; set; }
}

public class Employee3
{
    public string Name { get; init; } = default!;
}
```

And for the record DTO with empty constructor, it beats the purpose doesn't it? Once you create the object (using the empty constructor, you can't set the properties anymore if they are defined in an immutable way?

I guess I'm pointing this out to establish (coding practices) when to use records and when to use class in a consistent way as they are both capable of behaving in the same way (you can make a record mutable and a class immutable and vice-versa).

Personally I really like using the record type for 3 immutable fields or less because you can define it in a single line.

Reply:
Yes, we can make a record mutable and a class immutable, but that it's what I would say that beat the purpose.

I see the record as a value-object component, something that represents data in a given state and nothing more than that.
On the other hand, I see classes as a more object-oriented component where it defines more than data, having behaviors and/or a state.

The idea of having a record with an empty constructor in the way that is implemented is that we would be able to create a immutable record with flexibility, without having to define all fields

Since it's a DTO, I see it as a data representation, nothing more than a value-object, hence a record and not a class. 

Personally I think that records represent better DTOs in their concept than classes

But in practical terms, we can do it in either way and I think that this discussion is becoming bigger than it should be, so I will set those object as classes

And thank you very much for sharing your perspective, it is always enriching for the discussion to have different points of view.